asmcomp/linscan.ml

```diff
@@ -41,6 +45,25 @@ let rec insert_interval_sorted i = function
   | j :: _ as il when j.iend <= i.iend -> i :: il
   | j :: il -> j :: insert_interval_sorted i il
 
+let rec insert_int_sorted_uniq i = function
+    [] -> [i]
+  | j :: js as il ->
+      let c = Int.compare i j in
+      if c = 0
+      then il
+      else if c < 0 then i :: il
+      else j :: insert_int_sorted_uniq i js
+
+let rec release_expired_spilled ci pos = function
+    i :: il ->
+      if i.iend < pos then
+        let ss = match i.reg.loc with Stack(Local ss) -> ss | _ -> assert false in
+        ci.ci_free_slots <- insert_int_sorted_uniq ss ci.ci_free_slots;
+        release_expired_spilled ci pos il
+      else
+        i :: release_expired_spilled ci pos il
+  | [] -> []
```

Comment:
The other `released_*` functions have a common structure with `when i.iend >= pos` and `Internval.remove_expired_ranges`. Why does this function have a different structure?

Reply:
The difference in the `when` clause is just stylistic. As I undestand it, the `remove_expired_ranges` call is a space optimization: it removes the live ranges in each interval that end before the current position as they won't be used anymore; however, it makes the benchmark go noticeably slower:

| `N` | `trunk`  | `PR` | `PR`/`trunk` |
|--| ------------- | ------------- |--|
| 2000 | 0.363s | 0.582s  | 1.6 |
| 4000 | 1.266s | 2.347s  | 1.85 |
| 8000 | 4.884s | 10.244s  | 2.09 |

In fact, I would lean towards removing the _existing_ calls to `remove_expired_ranges` as it gives a noticeable speedup: it makes the version in this PR go faster than the existing version in `trunk`; but this is perhaps better left for a separate discussion.

Reply:
If the code is correct with and without `remove_expired_ranges` (I don't know), then I would prefer if the versions in the codebase were consistent with each other. Inconsistenties make development harder later because people spend time wondering whether identical things were intentionally done in a different way.

Reply:
Makes sense. I pushed a commit removing the calls to `remove_expired_ranges`: it gives a (very small, contrary to what I thought) speedup, but it is also simpler.

Reply:
After sleeping on it, I realized that these calls are needed to avoid a quadratic blowup in certain cases (when there is an interval consisting of very many separate live ranges). So I reverted this change and added a comment to the code instead explaining the asymmetry.